parsing/parser.mly

```diff
@@ -2045,28 +2054,30 @@ signature_item:
   MODULE
   ext = ext attrs1 = attributes
   name = mkrhs(module_name)
-  body = module_declaration_body
+  body = module_declaration_body(optional_atat_modalities_expr)
   attrs2 = post_item_attributes
   {
     let attrs = attrs1 @ attrs2 in
     let loc = make_loc $sloc in
     let docs = symbol_docs $sloc in
-    Md.mk name body ~attrs ~loc ~docs, ext
+    let mty, modalities = body in
+    Md.mk name mty ~attrs ~loc ~docs ~modalities, ext
   }
 ;
 
 (* The body (right-hand side) of a module declaration. *)
-module_declaration_body:
-    COLON mty = module_type
-      { mty }
+module_declaration_body(mm):
+    COLON mty = module_type mm = mm
+      { mty, mm }
   | EQUAL error
       { expecting $loc($1) ":" }
   | mkmty(
-      arg_and_pos = functor_arg body = module_declaration_body
+      arg_and_pos = functor_arg body = module_declaration_body(optional_atat_mode_expr)
         { let (_, arg) = arg_and_pos in
-          Pmty_functor(arg, body) }
+          let (ret, mret) = body in
+          Pmty_functor(arg, ret, mret) }
     )
-    { $1 }
+    { $1, [] }
```

Comment:
Should there be syntax to give a functor a mode? Something like this, to match functions?:

```ocaml
module (F @ modes) (Arg : Blah) : sig
end
```